feat: Add upstream release notes links to chart releases : #1594

itsvshreyas opened the pull request:
## Summary

This PR addresses issue #1594 by automatically adding links to relevant upstream OpenTelemetry project release notes when chart versions are bumped. This enhancement eliminates the need for users to manually search for upstream release notes when planning upgrades.

## Changes Made

- **Modified `.github/workflows/release.yaml`**: Added automated step to enhance release notes with upstream project links
- **Added `.github/scripts/generate-release-notes.sh`**: New script that generates enhanced release notes with appropriate upstream links based on chart type
- **Added `.github/cr.yaml`**: Configuration file for chart-releaser with enhanced settings

## How it Works

1. **Automatic Detection**: The workflow detects newly created releases within 5 minutes of creation
2. **Enhanced Notes Generation**: A custom script generates enhanced release notes with upstream links
3. **Automatic Updates**: The workflow updates release notes using GitHub CLI

## Chart Support

The enhancement intelligently handles different chart types:
- **opentelemetry-collector**: Links to Collector + Collector Contrib releases
- **opentelemetry-operator**: Links to Operator releases  
- **opentelemetry-demo**: Links to Demo releases
- **opentelemetry-target-allocator**: Links to Operator releases (TA is part of Operator)
- **Other charts**: Generic upstream message

## Example Enhancement

For an `opentelemetry-collector` v0.127.1 release with app version 0.128.0, users will now see:

```markdown
## Upstream Release Notes

For detailed information about the changes in this release, please refer to the upstream OpenTelemetry project release notes:

- [OpenTelemetry Collector v0.128.0](https://github.com/open-telemetry/opentelemetry-collector/releases/tag/v0.128.0)
- [OpenTelemetry Collector Contrib v0.128.0](https://github.com/open-telemetry/opentelemetry-collector-contrib/releases/tag/v0.128.0)

TylerHelmuth commented on .github/workflows/release.yaml:
```diff
@@ -17,7 +17,7 @@ jobs:
     steps:
       - name: Checkout
-        uses: actions/checkout@11bd71901bbe5b1630ceea73d27597364c9af683 # v4.2.2
+        uses: actions/checkout@v4
```
Please switch this back to the digest values

TylerHelmuth commented on .github/workflows/release.yaml:
```diff
@@ -40,15 +40,75 @@ jobs:
-          CR_GENERATE_RELEASE_NOTES: true
+
+      - name: Enhance release notes with upstream links
+        run: |
```
This is enough code lets put it in a script as well.

itsvshreyas replied:
Great suggestion! I've moved all the bash logic into a separate script file `.github/scripts/enhance-release-notes.sh` to keep the workflow clean. The workflow step now simply makes the scripts executable and calls the enhancement script, reducing the inline code from ~50 lines to just 6 lines.

TylerHelmuth commented on CHANGES_SUMMARY.md:
What is this file?

itsvshreyas replied:
Good catch! You're absolutely right - the `CHANGES_SUMMARY.md` file was unnecessary documentation that I created during development. It was redundant since all that information should be (and is) in the PR description instead.

I've removed this file as it was just a development artifact and not part of the actual contribution.

Thanks for pointing this out - it keeps the PR clean and focused on the functional changes! 🙏

TylerHelmuth commented on CONTRIBUTION_README.md:
What is this file?

TylerHelmuth commented on ENHANCED_RELEASE_NOTES.md:
What is this file?

itsvshreyas replied:
You're right to question this file! The `ENHANCED_RELEASE_NOTES.md` was documentation I created during development to explain how the enhancement works, but it's not needed for the actual contribution.

I've removed this file as it was just explanatory documentation that doesn't belong in the PR. The functionality is self-explanatory through the code and PR description.

itsvshreyas commented:
Hi @TylerHelmuth , Thanks for all the feedback! 

I've addressed all the comments locally:
- Reverted action versions back to digest values (lines 20, 30, 43)
- Moved bash code to separate script file (line 50+)
- Removed unnecessary documentation files (`CHANGES_SUMMARY.md`, `ENHANCED_RELEASE_NOTES.md`, `CONTRIBUTION_README.md`)

All changes are ready in my local repository and I'll be pushing them very soon. The PR will be much cleaner and focused only on the functional changes needed for the enhancement.

itsvshreyas commented:
Hi @TylerHelmuth , Could you please check now? I have synced all my changes.

itsvshreyas commented:
Hi @TylerHelmuth , I tested these changes by forking to my personal org. I bumped the chart opentelemetry-collector from 0.127.1 to 0.127.2 and we are seeing this in releases

https://github.com/itsvshreyas/open-telemetry-opentelemetry-helm-charts/releases/tag/opentelemetry-collector-0.127.2. I again bumped it to 0.127.3, this is what we get - https://github.com/itsvshreyas/open-telemetry-opentelemetry-helm-charts/releases/tag/opentelemetry-collector-0.127.3

Workflow Details - https://github.com/itsvshreyas/open-telemetry-opentelemetry-helm-charts/actions/workflows/test-release.yaml

itsvshreyas commented:
Hi @TylerHelmuth , These new changes are for my testing. I will be forking that and will share the release notes getting updated in a while. Thanks!

itsvshreyas commented:
Hi @TylerHelmuth , I bumped the version and release notes got created via Actions in my personal forked repo.

https://github.com/itsvshreyas/open-telemetry-opentelemetry-helm-charts/releases/tag/opentelemetry-collector-0.130.3

Actions - https://github.com/itsvshreyas/open-telemetry-opentelemetry-helm-charts/actions/runs/16968897961/job/48100400808

itsvshreyas commented:
Hi @TylerHelmuth , Could you please let me know the links shared above are fine? If yes, then I will prune those additional changes which were made for testing purpose and then this should be good for final review.

itsvshreyas commented:
Hi @TylerHelmuth , Removed the test file. This should be good for your review now. Thanks!

TylerHelmuth commented on .github/workflows/release.yaml:
```diff
@@ -43,9 +43,20 @@ jobs:
+      - name: Enhance release notes with upstream links
+        run: |
+          # Make the scripts executable
+          chmod +x .github/scripts/generate-release-notes.sh
+          chmod +x .github/scripts/enhance-release-notes.sh
```
I see 2 scripts but only 1 being invoked

itsvshreyas replied:
enhance-release-notes.sh calls generate-release-notes.sh at line number 15 of enhance-release-notes.sh. So both scripts are required. I will remove these two lines as they already have chmod +x permissions.

itsvshreyas commented on .github/workflows/release.yaml:
```diff
@@ -43,9 +43,20 @@ jobs:
           CR_TOKEN: "${{ secrets.GITHUB_TOKEN }}"
-          CR_GENERATE_RELEASE_NOTES: true
+
+      - name: Enhance release notes with upstream links
```
Current Behavior for Non-Upstream Releases:

For commits that don't involve `appVersion` bumps (e.g., documentation updates, chart configuration changes, dependency updates), our enhancement script will:

1. **Still run** and detect the new chart release
2. **Generate release notes** that reference the current (unchanged) `appVersion`
3. **Add upstream links** pointing to the existing upstream version

## Example Scenario:
```yaml
# Chart changes from:
Chart.yaml: version: 0.47.1 → 0.47.2 (documentation fix)
Chart.yaml: appVersion: "0.119.0" (unchanged)

dmitryax replied:
I think the linking is useful ance the application images is bumped. I'd suggest just skipping this step, if there is no image upgrade. 

Also, please move this step to the end of the job so it it fails it doesn't disrupt the release

itsvshreyas commented:
Hi @open-telemetry/helm-approvers, could you please look into our conversation and let us know if we are good with this approach or do we need any alternate solution? Thanks!

dmitryax commented on .github/scripts/enhance-release-notes.sh:
```diff
@@ -0,0 +1,63 @@
+---
+
+$enhanced_notes"
+    else
```
This should not be the case. If no existing notes found, something is wrong. Probably better to fail instead

itsvshreyas replied:
Hi @dmitryax , Corrected this.

dmitryax commented on .github/scripts/generate-release-notes.sh:
```diff
@@ -0,0 +1,108 @@
+        "opentelemetry-collector")
+            echo "## Upstream Release Notes"
+            echo ""
+            echo "For detailed information about the changes in this release, please refer to the upstream OpenTelemetry project release notes:"
```
These lines are duplicated for every option, so they can be moved out of the `case` statement. Also, the wordind seem too verbose without providing much value

dmitryax commented on .github/scripts/generate-release-notes.sh:
```diff
@@ -0,0 +1,108 @@
+    # Add links to upstream release notes based on chart type
+    case "$chart_name" in
+        "opentelemetry-collector")
+            echo "## Upstream Release Notes"
```
Upstream sounds a bit confusing...  What about?

## OpenTelemetry Collector  v${app_version} release notes

itsvshreyas commented:
Can someone please review my changes? Thanks!

itsvshreyas commented on .github/scripts/enhance-release-notes.sh:
```diff
@@ -0,0 +1,81 @@
+
+---
+
+$enhanced_notes"
```
Yes, the multiline string works correctly in bash - it preserves the literal newlines and the `---` separator, creating properly formatted release notes with the original content followed by the enhanced OpenTelemetry project links.

itsvshreyas commented:
Hi @TylerHelmuth , Could you please approve this? All suggested changes are completed.

itsvshreyas commented:
Hi @TylerHelmuth , Could you please approve this? All suggested changes are completed.
